scope.go

```diff
@@ -0,0 +1,603 @@
+package rcmgr
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+// resources tracks the current state of resource consumption
+type resources struct {
+	limit Limit
+
+	nconnsIn, nconnsOut     int
+	nstreamsIn, nstreamsOut int
+	nfd                     int
+
+	memory int64
+}
+
+// A ResourceScope can be a DAG, where a downstream node is not allowed to outlive an upstream node
+// (ie cannot call Done in the upstream node before the downstream node) and account for resources
+// using a linearized parent set.
+// A ResourceScope can be a txn scope, where it has a specific owner; txn scopes create a tree rooted
+// at the owner (which can be a DAG scope) and can outlive their parents -- this is important because
+// txn scopes are the main *user* interface for memory management, and the user may call
+// Done in a txn scope after the system has closed the root of the txn tree in some background
+// goroutine.
+// If we didn't make this distinction we would have a double release problem in that case.
+type ResourceScope struct {
+	sync.Mutex
+	done   bool
+	refCnt int
+
+	rc          resources
+	owner       *ResourceScope   // set in transaction scopes, which define trees
+	constraints []*ResourceScope // set in DAG scopes, it's the linearized parent set
```

Comment:
The rollback logic could be simplified by introducing a
```go
type constraints []*ResourceScope
```
that implements `network.ResourceScope`.

Reply:
sure, but it would still be the same code ;)

Reply:
as we discussed in sync, this doesn't seem worth right now; we may still refactr later.